File: src/services/ConversationService.ts
import { OpenAIService } from './OpenAIService';
import { ElevenLabsService } from './ElevenLabsService';
import { AudioService } from './AudioService';

export interface ConversationMessage {
  id: string;
  content: string;
  sender: 'user' | 'assistant';
  timestamp: Date;
  emotionalTone?: 'supportive' | 'encouraging' | 'corrective' | 'neutral';
}

export interface ConversationScenario {
  id: string;
  title: string;
  description: string;
  topic: string;
  goal: string;
  difficulty: number;
  quickResponses: string[];
}

export const CONVERSATION_SCENARIOS: ConversationScenario[] = [
  {
    id: '1',
    title: 'Social Greeting',
    description: 'Learn appropriate ways to greet others',
    topic: 'social_greeting',
    goal: 'Practice greeting people in different situations',
    difficulty: 1,
    quickResponses: [
      'Hi, how are you?',
      'Good morning!',
      'Nice to meet you',
      'Have a great day!',
    ],
  },
  {
    id: '2',
    title: 'Asking for Help',
    description: 'Learn to ask for help when needed',
    topic: 'asking_help',
    goal: 'Practice asking for assistance politely',
    difficulty: 2,
    quickResponses: [
      'Can you help me?',
      'I need assistance with...',
      'Could you please explain?',
      'Thank you for your help',
    ],
  },
  {
    id: '3',
    title: 'Expressing Feelings',
    description: 'Learn to express feelings appropriately',
    topic: 'expressing_feelings',
    goal: 'Practice sharing emotions in a healthy way',
    difficulty: 3,
    quickResponses: [
      'I feel happy when...',
      'I am frustrated because...',
      'This makes me feel...',
      'I appreciate that',
    ],
  },
  {
    id: '4',
    title: 'Daily Activities',
    description: 'Learn to discuss daily activities',
    topic: 'daily_activities',
    goal: 'Practice talking about routine and activities',
    difficulty: 1,
    quickResponses: [
      'Today I did...',
      'I enjoy...',
      'My favorite activity is...',
      'What do you like to do?',
    ],
  },
];

export class ConversationService {
  private messages: ConversationMessage[] = [];
  private currentScenario: ConversationScenario | null = null;
  private startTime: Date | null = null;
  private isListening: boolean = false;

  async startConversation(scenario: ConversationScenario): Promise<ConversationMessage> {
    this.currentScenario = scenario;
    this.messages = [];
    this.startTime = new Date();

    // Generate opening message
    const openingText = await this.generateOpeningMessage(scenario);

    const openingMessage: ConversationMessage = {
      id: this.generateMessageId(),
      content: openingText,
      sender: 'assistant',
      timestamp: new Date(),
      emotionalTone: 'supportive',
    };

    this.messages.push(openingMessage);

    // Play opening message via audio
    try {
      const audioUri = await ElevenLabsService.synthesizeSpeech(openingText, 'supportive');
      await AudioService.playSound(audioUri);
    } catch (error) {
      // Fallback to system TTS
      await AudioService.speak(openingText);
    }

    return openingMessage;
  }

  async processUserMessage(userInput: string): Promise<ConversationMessage> {
    // Add user message
    const userMessage: ConversationMessage = {
      id: this.generateMessageId(),
      content: userInput,
      sender: 'user',
      timestamp: new Date(),
    };
    this.messages.push(userMessage);

    // Generate AI response
    const conversationHistory = this.messages.map(msg => ({
      role: msg.sender === 'user' ? 'user' as const : 'assistant' as const,
      content: msg.content,
    }));

    const systemPrompt = this.buildSystemPrompt();
    const aiResponseText = await OpenAIService.chat(
      [{ role: 'system', content: systemPrompt }, ...conversationHistory],
      150
    );

    // Determine emotional tone
    const tone = this.determineEmotionalTone(aiResponseText);

    const assistantMessage: ConversationMessage = {
      id: this.generateMessageId(),
      content: aiResponseText,
      sender: 'assistant',
      timestamp: new Date(),
      emotionalTone: tone,
    };
    this.messages.push(assistantMessage);

    // Play response
    try {
      const audioUri = await ElevenLabsService.synthesizeSpeech(aiResponseText, tone);
      await AudioService.playSound(audioUri);
    } catch (error) {
      await AudioService.speak(aiResponseText);
    }

    return assistantMessage;
  }

  async endConversation(): Promise<{
    duration: number;
    messageCount: number;
    highlights: string;
    recommendations: string;
  }> {
    const endTime = new Date();
    const duration = this.startTime
      ? Math.floor((endTime.getTime() - this.startTime.getTime()) / 1000)
      : 0;

    const summary = await this.generateConversationSummary();

    return {
      duration,
      messageCount: this.messages.length,
      highlights: summary.highlights,
      recommendations: summary.recommendations,
    };
  }

  getMessages(): ConversationMessage[] {
    return this.messages;
  }

  getCurrentScenario(): ConversationScenario | null {
    return this.currentScenario;
  }

  private buildSystemPrompt(): string {
    if (!this.currentScenario) {
      return 'You are a supportive AI coach helping someone practice social conversations.';
    }

    return `You are AURA Coach, a friendly and supportive AI assistant helping someone practice social skills.

Current Scenario: ${this.currentScenario.title}
Topic: ${this.currentScenario.topic}
Goal: ${this.currentScenario.goal}
Description: ${this.currentScenario.description}

Your role:
- Provide natural, conversational responses
- Be encouraging and supportive
- Gently correct mistakes when needed
- Ask follow-up questions to keep the conversation going
- Stay on topic related to the scenario
- Keep responses concise (2-3 sentences maximum)
- Use simple, clear language
- Show empathy and understanding

Remember: You're helping someone practice social interactions in a safe, judgment-free environment.`;
  }

  private async generateOpeningMessage(scenario: ConversationScenario): Promise<string> {
    const greetings = {
      social_greeting: "Hi there! I'm so glad you're here to practice greetings with me. Let's start with a simple hello - how would you greet someone you just met?",
      asking_help: "Hello! Today we're going to practice asking for help. Remember, asking for help is a sign of strength, not weakness. What's something you might need help with?",
      expressing_feelings: "Hi! I'm here to help you practice expressing your feelings. It's important to share how we feel. Can you tell me about something that made you happy recently?",
      daily_activities: "Hey! Let's talk about our daily activities. I love hearing about what people do in their day. What did you do today that you enjoyed?",
    };

    return greetings[scenario.topic as keyof typeof greetings] || greetings.social_greeting;
  }

  private determineEmotionalTone(text: string): 'supportive' | 'encouraging' | 'corrective' | 'neutral' {
    const lowerText = text.toLowerCase();

    if (
      lowerText.includes('great job') ||
      lowerText.includes('excellent') ||
      lowerText.includes('wonderful')
    ) {
      return 'encouraging';
    }

    if (
      lowerText.includes('try') ||
      lowerText.includes('instead') ||
      lowerText.includes('better to')
    ) {
      return 'corrective';
    }

    if (
      lowerText.includes('understand') ||
      lowerText.includes('that\'s okay') ||
      lowerText.includes('it\'s normal')
    ) {
      return 'supportive';
    }

    return 'neutral';
  }

  private async generateConversationSummary(): Promise<{
    highlights: string;
    recommendations: string;
  }> {
    const conversationText = this.messages
      .map(msg => `${msg.sender}: ${msg.content}`)
      .join('\n');

    const summaryPrompt = `Based on this conversation practice session, provide:
1. A brief highlight (1-2 sentences) of what went well
2. A recommendation (1-2 sentences) for future practice

Conversation:
${conversationText}

Format your response as JSON:
{
  "highlights": "...",
  "recommendations": "..."
}`;

    try {
      const response = await OpenAIService.chat([
        { role: 'system', content: 'You are analyzing a social skills practice conversation.' },
        { role: 'user', content: summaryPrompt },
      ], 200);

      const parsed = JSON.parse(response);
      return {
        highlights: parsed.highlights || 'Great conversation practice!',
        recommendations: parsed.recommendations || 'Keep practicing regularly!',
      };
    } catch (error) {
      return {
        highlights: 'You engaged well in the conversation and stayed on topic.',
        recommendations: 'Continue practicing to build confidence in social situations.',
      };
    }
  }

  private generateMessageId(): string {
    return `msg-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
  }

  startListening() {
    this.isListening = true;
  }

  stopListening() {
    this.isListening = false;
  }

  getIsListening(): boolean {
    return this.isListening;
  }
}

File: legacy/swift-ios/Services/ConversationService.swift
import Foundation
import AVFoundation
import Speech

// MARK: - Real-time Conversation Service
class ConversationService: ObservableObject {
    static let shared = ConversationService()
    
    @Published var isActive = false
    @Published var currentConversation: Conversation?
    @Published var conversationHistory: [ConversationMessage] = []
    @Published var isListening = false
    @Published var isProcessing = false
    @Published var lastError: String?

    private let openAIService = OpenAIService.shared
    private let elevenLabsService = ElevenLabsService.shared
    private let audioEngine = AVAudioEngine()
    private let speechRecognizer = ConversationSpeechRecognitionManager()
    private let audioSession = AVAudioSession.sharedInstance()

    enum ConversationServiceError: Error {
        case microphoneDenied
    }
    
    // MARK: - Conversation Management
    func startConversation(scenario: ConversationScenario, userLevel: UserLevel) async throws {
        currentConversation = Conversation(
            id: UUID(),
            scenario: scenario,
            userLevel: userLevel,
            startTime: Date()
        )
        
        conversationHistory.removeAll()
        isActive = true
        
        let openingMessage: ConversationMessage
        do {
            openingMessage = try await generateOpeningMessage(for: scenario, userLevel: userLevel)
        } catch {
            openingMessage = fallbackOpeningMessage(for: scenario)
        }

        await addMessageToHistory(openingMessage)
        try? await elevenLabsService.playConversationResponse(openingMessage.content, tone: .supportive)
    }
    
    func endConversation() {
        isActive = false
        stopListening()
        currentConversation?.endTime = Date()
        
        // Generate conversation summary
        Task {
            await generateConversationSummary()
        }
    }
    
    // MARK: - Voice Interaction
    func startListening() async throws {
        guard !isListening else { return }
        try await configureAudioSession()

        isListening = true
        
        try await speechRecognizer.startRecording { [weak self] (recognizedText: String) in
            Task { @MainActor in
                if !recognizedText.isEmpty {
                    self?.isListening = false
                    await self?.processUserInput(recognizedText)
                }
            }
        }
    }
    
    func stopListening() {
        speechRecognizer.stopRecording()
        isListening = false
        try? audioSession.setActive(false, options: [.notifyOthersOnDeactivation])
    }
    
    func processUserMessage(_ message: String) async -> String {
        await processUserInput(message)
        // Return the AI's response from conversation history
        return conversationHistory.last?.content ?? "I'm sorry, I didn't understand that."
    }
    
    private func processUserInput(_ input: String) async {
        guard let conversation = currentConversation else { return }
        
        isProcessing = true
        defer { isProcessing = false }
        
        // Add user message to history
        let userMessage = ConversationMessage(
            id: UUID(),
            content: input,
            sender: .user,
            timestamp: Date(),
            emotionalTone: .neutral
        )
        await addMessageToHistory(userMessage)
        
        do {
            // Generate AI response
            let context = ConversationContext(
                userEmotionalState: "engaged", // TODO: Get from UserMonitoringService
                topic: conversation.scenario.topic,
                userLevel: conversation.userLevel.description
            )
            
            let response = try await openAIService.generateConversationResponse(userInput: input, context: context)

            let aiMessage = ConversationMessage(
                id: UUID(),
                content: response.text,
                sender: .assistant,
                timestamp: Date(),
                emotionalTone: response.emotionalTone
            )
            await addMessageToHistory(aiMessage)

            try? await elevenLabsService.playConversationResponse(response.text, tone: response.emotionalTone)
            
            // Auto-restart listening for continued conversation
            if isActive {
                try await startListening()
            }
            
        } catch {
            print("Error processing user input: \(error)")
            await MainActor.run { self.lastError = error.localizedDescription }
            let fallbackMessage = fallbackResponse(for: input)
            await addMessageToHistory(fallbackMessage)
            try? await elevenLabsService.playConversationResponse(fallbackMessage.content, tone: .supportive)
        }
    }

    private func configureAudioSession() async throws {
        let permissionGranted = await requestMicrophoneAccess()
        guard permissionGranted else { throw ConversationServiceError.microphoneDenied }

        try audioSession.setCategory(.playAndRecord, mode: .voiceChat, options: [.defaultToSpeaker, .allowBluetooth])
        try audioSession.setActive(true, options: [.notifyOthersOnDeactivation])
    }

    private func requestMicrophoneAccess() async -> Bool {
        await withCheckedContinuation { continuation in
            audioSession.requestRecordPermission { allowed in
                continuation.resume(returning: allowed)
            }
        }
    }
    
    // MARK: - Message Generation
    private func generateOpeningMessage(for scenario: ConversationScenario, userLevel: UserLevel) async throws -> ConversationMessage {
        let prompt = """
        Generate a friendly opening message for a conversation practice scenario.
        
        Scenario: \(scenario.description)
        User Level: \(userLevel.description)
        Goal: Help the user practice \(scenario.goal)
        
        Create a warm, encouraging opening that:
        1. Sets up the scenario naturally
        2. Uses language appropriate for the user's level
        3. Encourages the user to respond
        4. Feels like a real conversation starter
        
        Keep it under 30 words and make it conversational.
        """
        
        let response = try await openAIService.generateConversationResponse(
            userInput: prompt,
            context: ConversationContext(
                userEmotionalState: "ready",
                topic: scenario.topic,
                userLevel: userLevel.description
            )
        )
        
        return ConversationMessage(
            id: UUID(),
            content: response.text,
            sender: .assistant,
            timestamp: Date(),
            emotionalTone: response.emotionalTone
        )
    }
    
    private func generateConversationSummary() async {
        guard let conversation = currentConversation else { return }
        
        let messages = conversationHistory.map { "\($0.sender.rawValue): \($0.content)" }.joined(separator: "\n")
        
        let prompt = """
        Analyze this conversation practice session and provide feedback:
        
        Scenario: \(conversation.scenario.description)
        Duration: \(conversation.duration) seconds
        Messages: \(conversationHistory.count)
        
        Conversation:
        \(messages)
        
        Provide:
        1. What the user did well
        2. Areas for improvement
        3. Specific suggestions for next time
        4. Overall progress assessment
        
        Keep feedback positive and constructive.
        """
        
        do {
            let summaryResponse = try await openAIService.generateConversationResponse(
                userInput: prompt,
                context: ConversationContext(
                    userEmotionalState: "completed",
                    topic: "conversation_feedback",
                    userLevel: conversation.userLevel.description
                )
            )
            
            conversation.summary = summaryResponse.text
            
        } catch {
            print("Error generating conversation summary: \(error)")
        }
    }
    
    // MARK: - Helper Methods
    @MainActor
    private func addMessageToHistory(_ message: ConversationMessage) {
        conversationHistory.append(message)
        currentConversation?.messageCount += 1
    }
    
    private func handleConversationError(_ error: Error) async {
        let errorMessage = ConversationMessage(
            id: UUID(),
            content: "I'm sorry, I didn't catch that. Could you try saying that again?",
            sender: .assistant,
            timestamp: Date(),
            emotionalTone: .supportive
        )
        
        await addMessageToHistory(errorMessage)
        
        do {
            try await elevenLabsService.playConversationResponse(
                errorMessage.content,
                tone: .supportive
            )
        } catch {
            print("Error playing error message: \(error)")
        }
    }

    private func fallbackOpeningMessage(for scenario: ConversationScenario) -> ConversationMessage {
        let text: String
        switch scenario.topic {
        case "greetings":
            text = "Hi there! Let's practice greetings together. How would you like to start?"
        case "help_requests":
            text = "Hello! Tell me something you need help with and we'll practice asking together."
        case "emotions":
            text = "Let's talk about feelings. How are you feeling right now?"
        case "daily_life":
            text = "I'd love to hear about your day. What's one thing you did today?"
        default:
            text = "I'm ready to chat. What would you like to talk about today?"
        }
        return ConversationMessage(id: UUID(), content: text, sender: .assistant, timestamp: Date(), emotionalTone: .supportive)
    }

    private func fallbackResponse(for input: String) -> ConversationMessage {
        let lowercased = input.lowercased()
        let reply: String

        if lowercased.contains("hello") || lowercased.contains("hi") {
            reply = "Hello! It's nice to chat with you. How are you feeling right now?"
        } else if lowercased.contains("help") {
            reply = "Of course—what do you need help with? Let's practice asking together."
        } else if lowercased.contains("happy") || lowercased.contains("good") {
            reply = "That's wonderful! What made you feel happy today?"
        } else if lowercased.contains("sad") || lowercased.contains("upset") {
            reply = "I'm sorry you're feeling that way. Would you like to talk about it?"
        } else if lowercased.contains("thank") {
            reply = "You're welcome! I'm glad to practice with you."
        } else {
            reply = "That's interesting! Can you tell me a little more about that?"
        }

        return ConversationMessage(id: UUID(), content: reply, sender: .assistant, timestamp: Date(), emotionalTone: .supportive)
    }
    
    // MARK: - Conversation Scenarios
    func getAvailableScenarios(for userLevel: UserLevel) -> [ConversationScenario] {
        switch userLevel.level {
        case 1: // Beginner
            return [
                ConversationScenario(
                    id: "greeting",
                    title: "Saying Hello",
                    description: "Practice greeting someone new",
                    topic: "greetings",
                    goal: "Learn how to start a conversation",
                    difficulty: 1
                ),
                ConversationScenario(
                    id: "asking_help",
                    title: "Asking for Help",
                    description: "Practice asking for assistance",
                    topic: "help_requests",
                    goal: "Learn to ask for help politely",
                    difficulty: 1
                )
            ]
            
        case 2: // Intermediate
            return [
                ConversationScenario(
                    id: "small_talk",
                    title: "Making Small Talk",
                    description: "Practice casual conversation",
                    topic: "small_talk",
                    goal: "Learn to make friendly conversation",
                    difficulty: 2
                ),
                ConversationScenario(
                    id: "expressing_feelings",
                    title: "Sharing Feelings",
                    description: "Practice expressing emotions",
                    topic: "emotions",
                    goal: "Learn to communicate feelings",
                    difficulty: 2
                )
            ]
            
        case 3: // Advanced
            return [
                ConversationScenario(
                    id: "conflict_resolution",
                    title: "Resolving Disagreements",
                    description: "Practice handling conflicts",
                    topic: "conflict_resolution",
                    goal: "Learn to resolve disagreements calmly",
                    difficulty: 3
                ),
                ConversationScenario(
                    id: "job_interview",
                    title: "Job Interview",
                    description: "Practice professional conversations",
                    topic: "professional",
                    goal: "Learn workplace communication",
                    difficulty: 3
                )
            ]
            
        default:
            return getAvailableScenarios(for: UserLevel(level: 1, description: "Beginner"))
        }
    }
}

// MARK: - Speech Recognition Manager
class ConversationSpeechRecognitionManager: ObservableObject {
    private var audioEngine: AVAudioEngine?
    private var recognitionRequest: SFSpeechAudioBufferRecognitionRequest?
    private var recognitionTask: SFSpeechRecognitionTask?
    private let speechRecognizer = SFSpeechRecognizer()
    
    func startRecording(completion: @escaping (String) -> Void) async throws {
        // Request speech recognition permission
        let authStatus = await requestSpeechPermission()
        guard authStatus == .authorized else {
            throw ConversationError.speechPermissionDenied
        }
        
        audioEngine = AVAudioEngine()
        recognitionRequest = SFSpeechAudioBufferRecognitionRequest()
        
        guard let audioEngine = audioEngine,
              let recognitionRequest = recognitionRequest else {
            throw ConversationError.speechSetupFailed
        }
        
        let inputNode = audioEngine.inputNode
        recognitionRequest.shouldReportPartialResults = true
        
        recognitionTask = speechRecognizer?.recognitionTask(with: recognitionRequest) { result, error in
            if let result = result {
                let recognizedText = result.bestTranscription.formattedString
                if result.isFinal {
                    completion(recognizedText)
                }
            }
            
            if error != nil {
                self.stopRecording()
            }
        }
        
        let recordingFormat = inputNode.outputFormat(forBus: 0)
        inputNode.installTap(onBus: 0, bufferSize: 1024, format: recordingFormat) { buffer, _ in
            recognitionRequest.append(buffer)
        }
        
        audioEngine.prepare()
        try audioEngine.start()
    }
    
    func stopRecording() {
        audioEngine?.stop()
        audioEngine?.inputNode.removeTap(onBus: 0)
        recognitionRequest?.endAudio()
        recognitionTask?.cancel()
        
        audioEngine = nil
        recognitionRequest = nil
        recognitionTask = nil
    }
    
    private func requestSpeechPermission() async -> SFSpeechRecognizerAuthorizationStatus {
        return await withCheckedContinuation { continuation in
            SFSpeechRecognizer.requestAuthorization { status in
                continuation.resume(returning: status)
            }
        }
    }
}

// MARK: - Data Models
class Conversation: ObservableObject {
    let id: UUID
    let scenario: ConversationScenario
    let userLevel: UserLevel
    let startTime: Date
    var endTime: Date?
    var messageCount: Int = 0
    var summary: String?
    
    var duration: TimeInterval {
        return (endTime ?? Date()).timeIntervalSince(startTime)
    }
    
    init(id: UUID, scenario: ConversationScenario, userLevel: UserLevel, startTime: Date) {
        self.id = id
        self.scenario = scenario
        self.userLevel = userLevel
        self.startTime = startTime
    }
}

struct ConversationScenario: Equatable, Hashable {
    let id: String
    let title: String
    let description: String
    let topic: String
    let goal: String
    let difficulty: Int
}

struct ConversationMessage: Identifiable {
    let id: UUID
    let content: String
    let sender: MessageSender
    let timestamp: Date
    let emotionalTone: EmotionalTone
}

enum MessageSender: String {
    case user = "User"
    case assistant = "Assistant"
}

enum ConversationError: Error {
    case speechPermissionDenied
    case speechSetupFailed
    case processingError
}

// MARK: - Extensions
extension ConversationService {
    func exportConversationData() -> String {
        guard let conversation = currentConversation else { return "" }
        
        var export = """
        Conversation Practice Session
        Scenario: \(conversation.scenario.title)
        Date: \(conversation.startTime)
        Duration: \(Int(conversation.duration)) seconds
        Messages: \(conversation.messageCount)
        
        Conversation Log:
        """
        
        for message in conversationHistory {
            export += "\n[\(message.timestamp)] \(message.sender.rawValue): \(message.content)"
        }
        
        if let summary = conversation.summary {
            export += "\n\nSummary:\n\(summary)"
        }
        
        return export
    }
}
